SEP-TU-e-2024/website
# Leaderboard data

# Backend

- Added leaderboard and leaderboard entry serializer to create a fixed structure for how the data will be send to the front end. They are in a separate folder called data processing to separate them from serializers that are used on data directly from the data base.
- Sorting of the leaderboard is done in the leaderboard data processing, currently just on a hard coded scoring metric, but we can change this to utilize problem data. Simply call sort from least significant to most with optional reverse parameter to rank the entries.
- Added scripts with get requests for the leaderboard and leaderboard entry data, this is a get request since it is meant to retrieve data only and it allows the data to be cached by http protocols.
- Added urls with variables for leaderboard and leaderboard entry. These urls allow the specific data to be requested and linked to on other pages, we could even add a name to the urls to make it easier.

# Frontend
- Update leaderboard front end to utilize the data requests for the leaderboard.
- Add methods for creating rows, column names and column keys based on the entries and problem of the leaderboard object.
- Removed sorting from the leaderboard table, as the frontend should expect the backend to give a sorted list of entries.

# Testing
In order to test the new code in this PR consider the following steps.
- Pull code from GitHub.
- Run python manage.py migrate
- Run the server [python manage.py runserver]
- Add to your local database a specified problem, submissions linked to the specified problem, results with metric 'scoring_metric' linked to the submissions.
- Run the website [yarn vite]
- In the address of the website add /leaderboard/1, where the 1 is the problem id.
- A sorted leaderboard with data from the database should be shown.
